src/framework/ui/uiwidget.cpp

```diff
@@ -1248,22 +1303,32 @@ UIWidgetPtr UIWidget::getRootParent()
     return static_self_cast<UIWidget>();
 }
 
-UIWidgetPtr UIWidget::getChildAfter(const UIWidgetPtr& relativeChild)
+UIWidgetPtr UIWidget::getChildAfter(const UIWidgetPtr& relativeChild, bool visibleOnly)
 {
-    return relativeChild->m_childIndex == m_children.size() ?
-        nullptr : m_children[relativeChild->m_childIndex];
+    auto it = std::find(m_children.begin(), m_children.end(), relativeChild);
+    if (it != m_children.end() && ++it != m_children.end()) {
+        if (!visibleOnly || (*it)->isExplicitlyVisible())
+            return *it;
+    }
+    return nullptr;
 }
 
-UIWidgetPtr UIWidget::getChildBefore(const UIWidgetPtr& relativeChild)
+UIWidgetPtr UIWidget::getChildBefore(const UIWidgetPtr& relativeChild, bool visibleOnly)
 {
-    return relativeChild->m_childIndex <= 1 ? nullptr : m_children[relativeChild->m_childIndex - 2];
+    auto it = std::find(m_children.rbegin(), m_children.rend(), relativeChild);
+    if (it != m_children.rend() && ++it != m_children.rend()) {
+        if (!visibleOnly || (*it)->isExplicitlyVisible())
+            return *it;
+    }
+    return nullptr;
 }
 
-UIWidgetPtr UIWidget::getChildById(const std::string_view childId)
+UIWidgetPtr UIWidget::getChildById(const std::string_view& childId, bool visibleOnly)
 {
-    if (const auto it = m_childrenById.find(childId); it != m_childrenById.end())
-        return it->second;
-
+    for (const UIWidgetPtr& child : m_children) {
+        if (child->getId() == childId && (!visibleOnly || child->isExplicitlyVisible()))
+            return child;
+    }
     return nullptr;
 }
```

Comment:
The childIndex was designed to avoid looping all childrens, this breaks this behaviour.
We should create new functions for this e.g: 
getVisibleChildAfter
getVisibleChildBefore
getVisibleChildById

and to be honest I would just do it in Lua,
```
function UIWidget:getVisibleChildById(id)
  for _, child in ipairs(self:getChildren()) do
    if child:getId() == id and child:isExplicityVisible()  then
      return child
    end
  end
end
```